### PR TITLE
feat: apple support

### PR DESCRIPTION
**Pull request by pollend**

### CODE REVIEW

File: Source/Metal/DeviceMetal.mm [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+              if (IORegistryEntryGetParentEntry(entry, kIOServicePlane, &parent) == kIOReturnSuccess)
+              {
+                  m_Desc.adapterDesc.vendor = GetVendorFromID(GetEntryProperty(parent, CFSTR("vendor-id")));
+                  m_Desc.adapterDesc.deviceId = GetEntryProperty(parent, CFSTR("device-id"));
+                  IOObjectRelease(parent);
+              }
+              IOObjectRelease(entry);
+          }
+      }
+  }
```
Comment (pollend) [inline Source/Metal/DeviceMetal.mm:54]: something interesting, I guess apple doesn't report the device/vendor.
```
    "Apple": {
      "id": "0x106b",

      "_comment": [
        "Apple GPUs do not report a DeviceID via the Metal API, and as such the typical device",
        "pattern matching does not work for them. The recommended approach is to find the highest",
        "supported 'common' family supported and report it as the architecture.",
        "Examples: 'common-1', 'common-3'",
        "https://developer.apple.com/documentation/metal/gpu_devices_and_work_submission/detecting_gpu_features_and_metal_software_versions"
      ]
    },
```



```
  // No vendor id, device id for Apple GPUs
  if (strstr(m_Desc.adapterDesc.name, "Apple")) {
      m_Desc.adapterDesc.vendor = nri::Vendor::APPLE;
  } else {
      const uint64_t regID = [m_Device registryID];
      if (regID)
      {
          io_registry_entry_t entry = IOServiceGetMatchingService(kIOMasterPortDefault, IORegistryEntryIDMatching(regID));
          if (entry)
          {
              // That returned the IOGraphicsAccelerator nub. Its parent, then, is the actual PCI device.
              io_registry_entry_t parent;
              if (IORegistryEntryGetParentEntry(entry, kIOServicePlane, &parent) == kIOReturnSuccess)
              {
                  m_Desc.adapterDesc.vendor = GetVendorFromID(GetEntryProperty(parent, CFSTR("vendor-id")));
                  m_Desc.adapterDesc.deviceId = GetEntryProperty(parent, CFSTR("device-id"));
                  IOObjectRelease(parent);
              }
              IOObjectRelease(entry);
          }
      }
  }
```

Comment (vertver) [reply Source/Metal/DeviceMetal.mm:54]: I don't think it's a problem. Just do what Metal's layer says and just pass the device.

File: Include/NRIDescs.h [hunk 2]
```diff
@@ -1444,10 +1444,15 @@ NRI_ENUM
...
     D3D12,
     VK,
+    MTL,
 
     MAX_NUM
 );
 
+// defined in apple framework
+#undef INTEL
+#undef AMD
```
Comment (pollend) [inline Include/NRIDescs.h:1454]: looks like these are #defined in apple framework.

File: Source/Metal/ConversionMTL.h [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    MTLDataTypeSampler, // SAMPLER
+    MTLDataTypeNone,    // CONSTANT_BUFFER
+    MTLDataTypeTexture, // TEXTURE
+    MTLDataTypeNone,    // STORAGE_TEXTURE
+    MTLDataTypeStruct,  // BUFFER
+    MTLDataTypeStruct,  // STORAGE_BUFFER
+    MTLDataTypeArray,   // STRUCTURED_BUFFER
+    MTLDataTypeStruct,    // STORAGE_STRUCTURED_BUFFER
+    MTLDataTypePrimitiveAccelerationStructure     // ACCELERATION_STRUCTURE
+};
```
Comment (pollend) [inline Source/Metal/ConversionMTL.h:68]: not sure how this will work in practice.

File: Include/NRIDescs.h (hunk 2 above)
Comment (dzhdanNV) [reply Include/NRIDescs.h:1219]: I don't mind to add these undefs to the interface.

**Comment by pollend** [comment]
@dzhdanNV just an observation you nearly have a C interface any reason for the C++ code to abstract across that. you can also just union a bunch of stuff together for dx12/dx11/vulkan/metal. You also have all this logic to build tables of functions etc ... 

```
struct CmdBuffer {
   union {
      struct {...}dx11;
      struct {...}dx12;
      struct {...}metal;
   };
}

struct NriDevice {
   union {
      struct {...}dx11;
      struct {...}dx12;
      struct {...}metal;
   };
   int (*Method1)(NriDevice*...);
}
```

File: Source/Metal/CommandBufferMTL.mm [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+void CommandBufferMTL::DrawIndexed(const DrawIndexedDesc& drawIndexedDesc) {
+    id<MTLBuffer> indexBuffer = m_CurrentIndexCmd.m_Buffer->GetHandle();
+    [m_encoder drawIndexedPrimitives: m_CurrentPipeline->m_graphics.m_primitiveType
+                              indexCount:drawIndexedDesc.indexNum
+                               indexType: m_CurrentIndexCmd.m_Type
+                             indexBuffer: indexBuffer
+                       indexBufferOffset: m_CurrentIndexCmd.m_Offset];
+}
+void CommandBufferMTL::DrawIndirect(const Buffer& buffer, uint64_t offset, uint32_t drawNum, uint32_t stride, const Buffer* countBuffer, uint64_t countBufferOffset) {}
+void CommandBufferMTL::DrawIndexedIndirect(const Buffer& buffer, uint64_t offset, uint32_t drawNum, uint32_t stride, const Buffer* countBuffer, uint64_t countBufferOffset) {}
```
Comment (pollend) [inline Source/Metal/CommandBufferMTL.mm:92]: looks like there isn't a count buffer implement provided by the encoder. https://developer.apple.com/documentation/metal/mtlrendercommandencoder/render_pass_drawing_commands?language=objc

Comment (vertver) [reply Source/Metal/CommandBufferMTL.mm:92]: Ping me when you're done with the base implementation, I'll try to get samples to work.

Comment (vertver) [reply Source/Metal/CommandBufferMTL.mm:92]: Also, argument buffers can be used to emulate root constants/push constants. But yes, you can also use it for count buffers

**Comment by pollend** [comment]
been busy with something else, I'll have to catch up to this once i finish up what i want to do with this game project im working on.

**Comment by dzhdanNV** [comment]
Sure, no prob. And tons of thanks for working on native Apple support!

**Comment by dzhdanNV** [comment]
Thanks for the progress. Very appreciated.

**Comment by pollend** [comment]
so i guess you encode the data to an argument buffer. I guess to reproduce descriptor sets the pool  would allocate an argument buffer for n number of descriptor sets?

```
fragmentShader(       RasterizerData            in                 [[ stage_in ]],
               device FragmentShaderArguments & fragmentShaderArgs [[ buffer(AAPLFragmentBufferIndexArguments) ]])
```

**Comment by dzhdanNV** [comment]
To be honest I don't follow. Please, explain the problem and a solution. Is the question about indirect drawing or something else?

**Comment by vertver** [comment]
> I guess to reproduce descriptor sets the pool would allocate an argument buffer for n number of descriptor sets?
Correct. Metal requires to fill argument buffers even for non-indirect draw calls, so you need to store descriptors in it.

**Comment by pollend** [comment]
any clue how these properties line up with DescriptorRangeDesc?
https://developer.apple.com/documentation/metal/mtlargumentdescriptor?language=objc

```
for(size_t r = 0; r < descriptorSetDesc.rangeNum; r++) {
            MTLArgumentDescriptor* argDescriptor = [MTLArgumentDescriptor argumentDescriptor];
            const DescriptorRangeDesc* range = &descriptorSetDesc.ranges[r];
            argDescriptor.arrayLength = range->descriptorNum;
            argDescriptor.access = MTLBindingAccessReadWrite;
            argDescriptor.index = range->baseRegisterIndex;
            switch(range->descriptorType) {
                case DescriptorType::TEXTURE:
                    argDescriptor.dataType = MTLDataTypeTexture;
                    argDescriptor.textureType = MTLTextureType2D; // descriptor type does not have this
                    break;
                case DescriptorType::SAMPLER:
                    argDescriptor.dataType = MTLDataTypeSampler;
                    break;
                case DescriptorType::CONSTANT_BUFFER:
                case DescriptorType::STORAGE_TEXTURE:
                case DescriptorType::BUFFER:
                case DescriptorType::STORAGE_BUFFER:
                case DescriptorType::STRUCTURED_BUFFER:
                case DescriptorType::STORAGE_STRUCTURED_BUFFER:
                    argDescriptor.dataType = MTLDataTypeStruct;
                    break;
                case DescriptorType::ACCELERATION_STRUCTURE:
                    argDescriptor.dataType = MTLDataTypePrimitiveAccelerationStructure;
                    break;
            }
        }
```

**Comment by dzhdanNV** [comment]
An obvious suggestion: you can look at `MoltenVK` sources and work this way:
```
NRI => open NRI/VK => open corresponding place in MoltenVK => analyze, rearrange, copy & paste to MTL implementation
```
(looking at MTL docs...)

**Comment by pollend** [comment]
> An obvious suggestion: you can look at `MoltenVK` sources and work this way:
> 
> ```
> NRI => open NRI/VK => open corresponding place in MoltenVK => analyze, rearrange, copy & paste to MTL implementation
> ```
> 
> (looking at MTL docs...)

I've been looking at the MoltenVK source, thanks for the suggestion. umm, ill have to do a bit of research i don't want to have to extend `DescriptorType`. maybe there is a more ideal solution?

**Comment by dzhdanNV** [comment]
Why do you need to potentially extend `DescriptorType`?

**Comment by pollend** [comment]
> Why do you need to potentially extend `DescriptorType`? Because of this:
> 
> ```c++
>                 case DescriptorType::TEXTURE:
>                     argDescriptor.dataType = MTLDataTypeTexture;
>                     argDescriptor.textureType = MTLTextureType2D; // descriptor type does not have this
>                     break;
> ```
> 
> ?

https://developer.apple.com/documentation/metal/mtltexturetype

defines 1D/1DArray/2D/2dArray/Cube/CubeArray etc ...

**Comment by vertver** [comment]
You can do the same trick as here - https://github.com/NVIDIAGameWorks/NRI/blob/205cf0518fbef19b3e6a8506e359d276e68b90b3/Source/D3D11/DescriptorD3D11.hpp#L55. Just identify your texture type by texture parameters, that's all you need.

**Comment by pollend** [comment]
> You can do the same trick as here -
> 
> https://github.com/NVIDIAGameWorks/NRI/blob/205cf0518fbef19b3e6a8506e359d276e68b90b3/Source/D3D11/DescriptorD3D11.hpp#L55
> Just identify your texture type by texture parameters, that's all you need.

in what way?

**Comment by vertver** [comment]
AFAIK, MTLTextureDescriptor is a direct analogue of NriDescriptor, so it shouldn't be hard to identify the texture type when creating the descriptor. If I'm right, you can use your TextureViewType to identify the MTLTextureType.

**Comment by dzhdanNV** [comment]
The problem is that there are no `descriptors` in `DescriptorRangeDesc`. In other places - yes. In this and hopefully the only one place - no.

**Comment by vertver** [comment]
https://github.com/KhronosGroup/MoltenVK/blob/1d2ec638532268d076a550262df2e5e5e4fd2c9e/MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm#L1146C43-L1146C67
This type doesn't seem to be very important, as it's only used for size calculations.

File: Source/Metal/CommandBufferMTL.mm [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+void CommandBufferMTL::SetShadingRate(const ShadingRateDesc& shadingRateDesc) {
+
+    //[m_RendererEncoder sha]
+}
+
+void CommandBufferMTL::ClearAttachments(const ClearDesc* clearDescs, uint32_t clearDescNum, const Rect* rects, uint32_t rectNum) {
+  
+
+    Restore();
+}
```
Comment (pollend) [inline Source/Metal/CommandBufferMTL.mm:224]: there isn't a similar implementation in metal for this. umm I guess it would have to cache out a pipeline and render rectangles to the various layers then restore some of the command buffer state?

Comment (dzhdanNV) [reply Source/Metal/CommandBufferMTL.mm:224]: Probably you could look at MoltenVK (VK to Metal) for `vkCmdClearAttachments`.

P.S. (since I'm here) Really appreciate your work!

Comment (dzhdanNV) [reply Source/Metal/CommandBufferMTL.mm:224]: https://github.com/KhronosGroup/MoltenVK/blob/main/MoltenVK/MoltenVK/Commands/MVKCmdTransfer.mm#L1275

Comment (pollend) [reply Source/Metal/CommandBufferMTL.mm:224]: I did see that umm, that does make this complicated. i guess i have to cache all the pipeline state and submit it in one batch. I think you can only have one encoder at any one time for a cmd buffer. does make the data structure a bit fat.

Comment (dzhdanNV) [reply Source/Metal/CommandBufferMTL.mm]: You can try suggesting API changes simplifying Metal implementation. Very undesired to be honest :)

Comment (pollend) [reply Source/Metal/CommandBufferMTL.mm]: at the moment im tracking any state for the vieport and only start a rendering encoder when a pipeline is set. you can have a look at the changes.

File: Source/Metal/CommandBufferMTL.mm [hunk 6]
```diff
@@ -0,0 +1,620 @@
...
+                      ];
+}
+void CommandBufferMTL::Dispatch(const DispatchDesc& dispatchDesc) {
+    InsertBarriers();
+    
+    MTLSize threadgroupCount = MTLSizeMake(dispatchDesc.x, dispatchDesc.y, dispatchDesc.z);
+//    MTLSize threadPerThreadGroup = MTLSizeMake(dispatchDesc.x, dispatchDesc.y, dispatchDesc.z);
+//    [m_ComputeEncoder
+//     dispatchThreadgroups: threadgroupCount
+//     threadsPerThreadgroup: threadPerThreadGroup];
```
Comment (pollend) [inline Source/Metal/CommandBufferMTL.mm:424]: you can pass the number of `dispatchThreadgroups` and `threadsPerThreadgroup`

**Comment by pollend** [comment]
is there any differences with managing memory with objective c. assignment rules and how things are freed?

File: Source/Metal/CommandBufferMTL.mm (hunk 6 above)
Comment (pollend) [reply Source/Metal/CommandBufferMTL.mm:421]: any clue how this would be managed?

File: Source/Metal/CommandBufferMTL.mm (hunk 4 above)
Comment (pollend) [reply Source/Metal/CommandBufferMTL.mm]: @vertver I think i have the core pieces in place still not sure about the overall implementation.

File: Source/Metal/DeviceMTL.mm [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+    id<MTLLibrary> mtlLib = [dev newLibraryWithSource: src
+                                                   options: options
+                                                     error: &err];    // temp retain
+    NSCAssert(err,  @"Failed to load Metal shader library %@", err );
+    
+    return [mtlLib newFunctionWithName: method];
+}
+
+
+id<MTLRenderPipelineState> DeviceMTL::GetClearPipeline(ClearDesc* desc, size_t numFormats) {
```
Comment (pollend) [inline Source/Metal/DeviceMTL.mm:93]: for this chunk of logic I need a way to cache pipelines for clearing. Any clue how this might be implemented for this portion of the code.

File: Source/Metal/CommandBufferMTL.mm (hunk 6 above)
Comment (dzhdanNV) [reply Source/Metal/CommandBufferMTL.mm:429]: MoltenVK pulls `threadsPerThreadgroup` from shaders and stores in pipelines. This value becomes available when a pipeline gets bound.

File: Source/Metal/DeviceMTL.mm (hunk 7 above)
Comment (dzhdanNV) [reply Source/Metal/DeviceMTL.mm:93]: (sorry, hard to comment on other questions due to lack of expertise)

Comment (pollend) [reply Source/Metal/DeviceMTL.mm]: i see, ill have to do my due diligence. im somewhat close to getting something working. might take a bit longer before i finalize anything.

File: Source/Metal/CommandBufferMTL.mm [hunk 8]
```diff
@@ -0,0 +1,689 @@
...
+void CommandBufferMTL::SetDescriptorPool(const DescriptorPool& descriptorPool) {
+    
+}
+
+
+void CommandBufferMTL::InsertBarriers() {
+    if(m_RendererEncoder) {
+        if (m_barrierFlags & BarrierBits::BARRIER_FLAG_BUFFERS)
+        {
+            [m_RendererEncoder memoryBarrierWithScope:MTLBarrierScopeBuffers
```
Comment (K0bin) [inline Source/Metal/CommandBufferMTL.mm:179]: I don't think this is sufficient at all.

Metal memory barriers only impact the pass they are in: https://developer.apple.com/videos/play/wwdc2022/10101/?time=1495
So it could still overlap with a different pass in the same command buffer or with another command buffer.

Meanwhile Vulkan barriers (and I assume the NRI barriers that look very similar) go beyond that and impact the whole queue.

**Review by K0bin — CHANGES_REQUESTED** [review]

**Comment by MERL10N** [comment]
Hey, fellow Mac user dabbling with Metal 4 and metal-cpp for that matter. Do you think it is a good idea if the metal code done so far can be converted to metal-cpp instead? It would be a waste to start all over again so with the current progress, I can assist with the migration. 

As to whether or not we should jump right into metal 4, it depends on everyone involved. Metal 3 still supports Hardware Raytracing through MTLAccelerationStructures. In addition, with Residency Sets Introduced in Metal 3.2 and the use of Argument Buffers, the metal backend will still be pretty performant. Metal 4 provides nice things like MetalFX Denoising and Neural Rendering Support (like Neural Texture Compression), but it would require a more substantial rewrite due to changes in the core API. 



Edit: Metal 4 also includes support for MetalFX Frame Interpolation (Apple's version of DLSS frame gen) and for M5 and up, support for Neural Temporal Upscaling.

**Comment by pollend** [comment]
> Hey, fellow Mac user dabbling with Metal 4 and metal-cpp for that matter. Do you think it is a good idea if the metal code done so far can be converted to metal-cpp instead? It would be a waste to start all over again so with the current progress, I can assist with the migration.
> 
> As to whether or not we should jump right into metal 4, it depends on everyone involved. Metal 3 still supports Hardware Raytracing through MTLAccelerationStructures. In addition, with Residency Sets Introduced in Metal 3.2 and the use of Argument Buffers, the metal backend will still be pretty performant. Metal 4 provides nice things like MetalFX Denoising and Neural Rendering Support (like Neural Texture Compression), but it would require a more substantial rewrite due to changes in the core API.
> 
> Edit: Metal 4 also includes support for MetalFX Frame Interpolation (Apple's version of DLSS frame gen) and for M5 and up, support for Neural Temporal Upscaling.

I've been working on my own little abstraction layer: [flying-swallow/rhi-zig](https://github.com/flying-swallow/rhi-zig) takes a lot of ideas for NRI. It's for zig but I don't see why it can't be rewritten for C++. I wanted an agnostic layer that doesn't really have a strong opinion so it's a lot more verbose. There are some slightly higher-level primitives that are more opinionated. seems like it will always be hard to support more than one graphics API ... webgl uses a FSM and webgpu has something like a command buffer but there is complication to deal with there. parts of the API are basically stubbed because metal doesn't have the same in either DirectX or Vulkan etc ... The ways metal does command buffers is uniquely just metal etc ...  Probably a better direction not to hide the details of the abstraction since all these graphics API have their own warts to deal with when you want to incorporate them.

**Comment by MERL10N** [comment]
> I've been working on my own little abstraction layer: [https://github.com/[flying-swallow/rhi-zig](https://github.com/flying-swallow/rhi-zig)](https://github.com/%5Bflying-swallow/rhi-zig%5D(https://github.com/flying-swallow/rhi-zig)) takes a lot of ideas for NRI. It's for zig but I don't see why it can't be rewritten for C++. I wanted an agnostic layer that doesn't really have a strong opinion so it's a lot more verbose. There are some slightly higher-level primitives that are more opinionated. seems like it will always be hard to support more than one graphics API ... webgl uses a FSM and webgpu has something like a command buffer but there is complication to deal with there. parts of the API are basically stubbed because metal doesn't have the same in either DirectX or Vulkan etc ... The ways metal does command buffers is uniquely just metal etc ... Probably a better direction not to hide the details of the abstraction since all these graphics API have their own warts to deal with when you want to incorporate them.

That's really cool! And yeah, Metal has its own quirks when it comes to getting it to work with RHIs designed for Vulkan and DX12 in mind. With that being said, I have been examining the code base for NRI and how other Graphics APIs are integrated into it, and I came to the conclusion that supporting Metal 4 from the get go is actually a better fit for NRI since Metal 4 is more aligned with Vulkan and DX12. For instance, Metal 4 now has Command Allocators and Metal 4's Argument Table are closer to Vulkan's Descriptor Sets than the resource binding methods we were used to with Metal 3. In short, there should be less stubbing to worry about with Metal 4 compared to Metal 3. But of course, I will need to start getting my hands dirty to see if that really is the case.

**Comment by dzhdanNV** [comment]
Design suggestion after reviewing the current Objective-C++/legacy Metal prototype:

### Recommendation

Use **C++ through Apple's [metal-cpp](https://developer.apple.com/metal/cpp/)** and target the **[Metal 4 core API](https://developer.apple.com/documentation/Metal/understanding-the-metal-4-core-api)** for the initial native backend.

This best matches the goals of reducing backend-specific code and staying close to Vulkan/D3D12:

| NRI | Metal 4 |
| --- | --- |
| `Queue` | `MTL4CommandQueue` |
| `CommandAllocator` | `MTL4CommandAllocator` |
| `CommandBuffer` | reusable `MTL4CommandBuffer` |
| descriptor-set storage | buffer-backed argument buffers |
| stage resource bindings | `MTL4ArgumentTable` |
| `CmdBarrier` | producer/consumer command barriers |
| residency | `MTLResidencySet` |
| pipeline compilation | `MTL4Compiler` |

#### Why metal-cpp

- It is Apple's official, header-only, one-to-one C++ interface with no measurable call overhead.
- The backend can retain NRI's existing `.h` / `.hpp` / `Impl*.cpp` organization.
- Only one implementation TU needs `NS_PRIVATE_IMPLEMENTATION`, `CA_PRIVATE_IMPLEMENTATION`, and `MTL_PRIVATE_IMPLEMENTATION`.
- `NS::SharedPtr<T>` can provide consistent ownership.
- The existing `MetalWindow::caMetalLayer` already supplies a `CAMetalLayer*`, so an Objective-C window shim should not be necessary.

#### Why Metal 4

Metal 4 introduces the explicit command allocator, reusable queue-independent command buffer, argument-table, residency, and barrier abstractions that NRI already models. The legacy Metal command API requires more emulation.

The `MTL4` command types are independent replacements rather than extensions of the Metal 3 types, so supporting both in one initial backend would add substantial branching and duplicate implementation. I suggest a Metal 4-only first version, using ordinary shared Metal types where Metal 4 did not introduce replacements.

The trade-off is a baseline of Apple silicon with Metal 4 support—M1/A14 or newer on macOS/iOS 26 or newer. Intel Macs and older OS versions would require a separate Metal 3 compatibility path later.

### Shader path

Prefer:

```text
HLSL -> DXIL -> Metal Shader Converter -> Metal IR/metallib
```

Apple's [Metal Shader Converter](https://developer.apple.com/metal/shader-converter/) supports DXIL and an explicit-layout mode. Deriving that layout from `PipelineLayoutDesc` would preserve NRI descriptor sets, root descriptors, and root constants while reusing the D3D12 shader artifacts.

Descriptor-set contents should live in argument buffers. `MTL4ArgumentTable` should bind those buffers and direct resources to shader stages; it should not be treated as the descriptor-pool allocator itself. Converted shaders should be cached by bytecode plus pipeline-layout signature, with direct metallib ingestion potentially added later for fully offline workflows.

In short: **metal-cpp + Metal 4 + DXIL/Metal Shader Converter** is the cleanest initial architecture for the stated goals.

**Comment by dzhdanNV** [comment]
(The Evil Me had tried to edit the original message but I stopped him ^_^)

**Comment by MERL10N** [comment]
> (The Evil Me had tried to edit the original message but I stopped him ^_^)

As a bored college student with nothing better to do over the summer (I still have a month to go before classes begin), working on the Metal 4 backend will help keep me busy. Let's see if I can get the clear window and triangle sample (From NRISamples) working on Metal 4 at the very least. 😄
